File: 프로그래머스/1/389478. 택배 상자 꺼내기/택배 상자 꺼내기.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

int solution(int n, int w, int num) {
    int h = (num - 1) / w;      
    
    int cnt = 0;
    while (n >= num){
        if (h % 2 == 0){
            int col =  w - 1 - ((num - 1) % w);
            num += col * 2 + 1;
        }
        else{
            int col = (num - 1) % w;
            num += (w - col) * 2 - 1;
        }
        cnt++;
        // cout << num << " ";
    }

    return cnt;
}

File: 프로그래머스/1/389478. 택배 상자 꺼내기/택배 상자 꺼내기.py
def idx_x(n, w, h):
    if h % 2 == 1:
        return (w - (n % w)) - 1
    return n % w 

def solution(n, w, num):
    
    n_h = (n - 1) // w
    n_w = idx_x(n-1, w, n_h)
    num_h = (num - 1) // w
    num_w = idx_x(num-1, w, num_h)
    
    # print(n_h, num_h, n_w, num_w)
    if n_h % 2 == 1 and n_w <= num_w:
        return n_h - num_h + 1
    elif n_h % 2 == 0 and n_w >= num_w:
        return n_h - num_h + 1
    else:
        return n_h - num_h